app/api/logging/audit.py

```diff
@@ -19,54 +19,93 @@
+    #     that are marked as restricted, and members that may allow bypassing imports.
+    #
+    #     Why we aren't logging:
+    #     The logging module itself accesses frame.f_code.co_filename
+    #     (see https://github.com/python/cpython/blob/bd7903967cd2a19ebc842dd1cce75f60a18aef02/Lib/logging/__init__.py#L202)
+    #     which calls object.__getattr__ under the hood
+    #     (see https://github.com/python/cpython/blob/main/Objects/frameobject.c#L89-L95)
+    #
+
+    EVENTS_TO_LOG = {
+        # Detect dynamic execution of code objects. This only occurs for explicit
+        # calls, and is not raised for normal function invocation.
+        "exec": ("code_object",),
+        # Detect when a file is about to be opened. path and mode are the usual
+        # parameters to open if available, while flags is provided instead of
+        # mode in some cases.
+        "open": ("path", "mode", "flags"),
+        # Detect access to network resources. The address is unmodified from the original call.
+        "socket.connect": ("socket", "address"),
+        # Detect when new audit hooks are being added.
+        "sys.addaudithook": (),
+        # Detects URL requests.
+        # Don't log data or headers because they may contain sensitive information.
+        "urllib.Request": ("url", "_", "_", "method"),
+    }
+
+    if event_name not in EVENTS_TO_LOG:
+        return
 
+    arg_names = EVENTS_TO_LOG[event_name]
+    log_audit_event(event_name, args, arg_names)
 
-IGNORE_AUDIT_EVENTS = {
-    "builtins.id",
-    "code.__new__",
-    "compile",
-    "exec",
-    "import",
-    "marshal.loads",
-    "object.__getattr__",
-    "object.__setattr__",
-    "os.listdir",
-    "os.scandir",
-    "os.walk",
-    "socket.__new__",
-    "sys._current_frames",
-    "sys._getframe",
-    "sys.settrace",  # interferes with PyCharm debugger
-}
 
+def log_audit_event(event_name: str, args: Sequence[Any], arg_names: Sequence[str]) -> None:
+    """Log a message but only log recently repeated messages at intervals."""
+    extra = {
+        f"audit.args.{arg_name}": arg for arg_name, arg in zip(arg_names, args) if arg_name != "_"
+    }
 
-def audit_hook(event_name: str, args: tuple[Any, ...]) -> None:
-    if event_name in IGNORE_AUDIT_EVENTS:
-        return
-    if event_name == "open" and isinstance(args[0], str) and "/__pycache__/" in args[0]:
-        # Python has a high rate of these events in normal operation.
-        return
-    if event_name == "os.chmod" and type(args[0]) is int:
-        # Gunicorn generates a high volume of these events in normal operation (see workertmp.py)
-        return
-    if event_name == "open" and isinstance(args[0], str) and "/pytz/" in args[0]:
-        # The pytz module generates a high volume of these events in normal operation.
+    key = (event_name, repr(args))
+    if key not in audit_message_count:
+        count = 1
+    else:
+        count = audit_message_count[key] + 1
```

Comment:
Don't need the `if` as `LeastRecentlyUsedDict` has logic to return a default of 0 if the key doesn't exist.

https://github.com/navapbc/template-application-flask/blob/802127b73475167feefb1e8541d1b89ce246fc3b/app/api/util/collections/dict.py#L27-L30

```suggestion
    count = audit_message_count[key] + 1
```

Reply:
oh hm, that's a little confusing that the dict returns a `0`, is this not a generic dict type or is this something that specifically has type `Mapping[str, int]`?

Reply:
@lorenyu it's a customized class that inherits from `OrderedDict` but has LRU and default value behaviour built-in. The idea is that we don't want every audit log entry ever seen in it, or it could keep growing without limit. So it only keeps the most recently seen ones (128 of them). See https://github.com/navapbc/template-application-flask/blob/main/app/api/util/collections/dict.py.

Reply:
@jamesbursa oh i see, didn't realize it was a custom class, the name looked like it could have been a standard library class.

if this the only place that uses it, then I'll move the class into this module so we can minimize generic utils that actually have very specific use cases. created a ticket for that https://github.com/navapbc/template-application-flask/issues/121